pkg/auth/providers/azure/azure_provider.go

```diff
@@ -0,0 +1,261 @@
+		if err != nil {
+			return v3.Principal{}, nil, nil, err
+		}
+	}
+
+	azureClient, err := newClientCode(azureCredential.Code, config)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+
+	providerInfo, err := ap.createProviderInfo(azureClient)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+
+	oid, err := parseJWTforField(azureClient.accessToken(), "oid")
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+
+	user, err := azureClient.userClient.Get(context.Background(), oid)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+
+	userPrincipal := ap.userToPrincipal(user)
+
+	var mem bool
+	params := graphrbac.UserGetMemberGroupsParameters{
+		SecurityEnabledOnly: &mem,
+	}
+
+	userGroups, err := azureClient.userClient.GetMemberGroups(context.Background(), *user.ObjectID, params)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+
+	groupPrincipals, err := ap.userGroupsToPrincipals(azureClient, userGroups)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+
+	testAllowedPrincipals := config.AllowedPrincipalIDs
+	if test && config.AccessMode == "restricted" {
+		testAllowedPrincipals = append(testAllowedPrincipals, userPrincipal.Name)
+	}
+
+	allowed, err := ap.userMGR.CheckAccess(config.AccessMode, testAllowedPrincipals, userPrincipal, groupPrincipals)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+	if !allowed {
+		return v3.Principal{}, nil, nil, httperror.NewAPIError(httperror.Unauthorized, "unauthorized")
+	}
+
+	return userPrincipal, groupPrincipals, providerInfo, nil
+}
+
+func (ap *azureProvider) saveAzureConfigK8s(config *v3.AzureADConfig) error {
+	storedAzureConfig, err := ap.getAzureConfigK8s()
+	if err != nil {
+		return err
+	}
+	config.APIVersion = "management.cattle.io/v3"
+	config.Kind = v3.AuthConfigGroupVersionKind.Kind
+	config.Type = client.AzureADConfigType
+	config.ObjectMeta = storedAzureConfig.ObjectMeta
+
+	logrus.Debugf("updating AzureADConfig")
+	_, err = ap.authConfigs.ObjectClient().Update(config.ObjectMeta.Name, config)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ap *azureProvider) getAzureConfigK8s() (*v3.AzureADConfig, error) {
+	authConfigObj, err := ap.authConfigs.ObjectClient().UnstructuredClient().Get(Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve AzureADConfig, error: %v", err)
+	}
+
+	u, ok := authConfigObj.(runtime.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("failed to retrieve AzureADConfig, cannot read k8s Unstructured data")
+	}
+	storedAzureADConfigMap := u.UnstructuredContent()
+
+	storedAzureADConfig := &v3.AzureADConfig{}
+	mapstructure.Decode(storedAzureADConfigMap, storedAzureADConfig)
+
+	metadataMap, ok := storedAzureADConfigMap["metadata"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("failed to retrieve AzureADConfig metadata, cannot read k8s Unstructured data")
+	}
+
+	typemeta := &metav1.ObjectMeta{}
```

Comment:
actually a naming mistake in the githubProvider. variable should be called objectMeta, not typemeta